src/constructors/tessellation.c

```diff
@@ -0,0 +1,304 @@
+    igraph_integer_t first_threshold = fmin(size_y - 1, size_z - 1);
+    igraph_integer_t second_threshold = fmax(size_y - 1, size_z - 1);
+    igraph_integer_t sgn_flag = size_y < size_z ? 0 : -1;
+
+    for (i = 0; i < row_count; i++)
+    {
+        VECTOR(row_lengths_vector)[i] = row_length;
+        VECTOR(row_start_vector)[i] = row_start;
+
+        if (i < first_threshold)
+        {
+            row_length++;
+            row_start--;
+        }
+        else if (i < second_threshold)
+        {
+            row_start += sgn_flag;
+        }
+        else
+        {
+            row_length--;
+        }
+    }
+
+    IGRAPH_CHECK(triangle_lattice(graph, directed, mutual, row_lengths_vector, row_start_vector));
+
+    igraph_vector_int_destroy(&row_lengths_vector);
+    IGRAPH_FINALLY_CLEAN(1);
+    igraph_vector_int_destroy(&row_start_vector);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \ingroup generators
+ * \function igraph_triangle_lattice
+ * \brief A triangle lattice with the given shape.
+ *
+ * Creates a triangle lattice whose vertices have the form (i, j) for non-negative integers i and j
+ * and (i, j) is generally connected with (i + 1, j), (i, j + 1), and (i - 1, j + 1).
+ *
+ * </para><para>
+ * The vertices of the resulting graph are ordered lexicographically with the 2nd coordinate being
+ *  more significant, e.g., (i, j) < (i + 1, j) and (i + 1, j) < (i, j + 1)
+ *
+ * \param graph An uninitialized graph object.
+ * \param dims Integer vector, defines the shape of the lattice. (Below the "edge length"s are in terms of graph theoretical path lengths.)
+ *        If the size of \p dims is 1 the resulting  lattice has a triangular shape with "edge length" dims[0].
+ *        If the size of \p dims is 2 the resulting lattice has a "quasi rectangular" shape with "edge length"s dims[0] and dims[1].
+ *        If the size of \p dims is 3 the resulting  lattice has a hexagonal shape with "edge length"s dims[0], dims[1] and dims[2].
+ *          (see https://github.com/igraph/igraph/issues/1842  for lattice visualizations)
+ * \param directed Boolean, whether to create a directed graph.
+ *        If the \c mutual argument is  not set to true,
+ *        edges will be directed from lower-index vertices towards
+ *        higher-index ones.
+ * \param mutual Boolean, if the graph is directed this gives whether
+ *        to create all connections as mutual.
+ * \return Error code:
+ *         \c IGRAPH_EINVAL: The size of \p dims must be either 1, 2, or 3 with all the components
+ *         at least 1.
+ *
+ * Time complexity:  O(|V|), where |V| is the number of vertices in the generated graph.
+ *
+ */
+igraph_error_t igraph_triangle_lattice(igraph_t *graph, const igraph_vector_int_t *dims, igraph_bool_t directed, igraph_bool_t mutual)
+{
+    if (dims->stor_begin == NULL)
+    {
+        IGRAPH_ERRORF("dims vector must be initialized  (%" IGRAPH_PRId ").", IGRAPH_EINVAL, (igraph_integer_t) 0);
```

Comment:
I am throwing an error in this way since using `IGRAPH_ERROR` didn't work for 
`IGRAPH_ASSERT(igraph_triangle_lattice(&graph, &dimvector, true, true) == IGRAPH_EINVAL);` when `dimvector` was not initialized.